cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java

```diff
@@ -134,15 +151,37 @@ protected final String loadMessageId() {
   @Nullable
   @Override
   protected String processMessages(Iterable<ImmutablePair<String, T>> messages) throws Exception {
-    MessageTrackingIterator iterator;
 
     // Process the notifications and record the message id of where the processing is up to.
     // 90% of the tx timeout is .9 * 1000 * txTimeoutSeconds = 900 * txTimeoutSeconds
     long timeBoundMillis = 900L * txTimeoutSeconds;
-    iterator = TransactionRunners.run(getTransactionRunner(), context -> {
-      TimeBoundIterator<ImmutablePair<String, T>> timeBoundMessages = new TimeBoundIterator<>(messages.iterator(),
-                                                                                              timeBoundMillis);
-      MessageTrackingIterator trackingIterator = new MessageTrackingIterator(timeBoundMessages);
+    TimeBoundIterator<ImmutablePair<String, T>> timeBoundMessages = new TimeBoundIterator<>(messages.iterator(),
+                                                                                            timeBoundMillis);
+
+    if (this.txSize >= this.getFetchSize()) {
```

Comment:
this still will result in extra memory usage for some scenarios, it would be better to do something more like:

```
while (timeBoundMessages.hasNext()) {
  LimitIterator iterator = new LimitIterator(timeBoundMessages, this.txSize);
  try {
    lastMessageId = doProcess(iterator);
  } ...
}
```

where the LimitIterator is similar to the one we have in NoSqlStructuredTable.

Reply:
The main challenge is that since txn may be retried, we may have to go over same entries several times. That is why originally, we are using Iterable, and we can call iterator on it inside the txn. This allows us to consume same messages several times inside txn if txn is retried. 
Let me see how I can add this additional logic to LimitIterator

Reply:
The problem with current approach is that you can't stop in the middle of the batch due to the fact that time is up. As soon as you read the batch, you'll process it to the end even if time is over.

Reply:
Done.

Reply:
Agree. Most of the complexity was coming from the fact that we wanted to retain both logics. Changed the PR, and it's much simpler.